Babich_8_home_work

Yandeks opened the pull request:
VS2017 выдает странную ошибку, пишут, что это ее проблемы, а не проблемы кода. Поэтому я решил проверить код здесь.

Yandeks committed:
Babich_4_home_work

Yandeks committed:
Update Array.h

Yandeks committed:
Update Array.cpp

Yandeks committed:
Update main_Array.cpp

Yandeks committed:
Update Array.cpp

Yandeks committed:
Update Array.cpp

Yandeks committed:
Update Array.h

Yandeks committed:
Update Array.h

Yandeks committed:
Update Array.cpp

Yandeks committed:
2/3

Yandeks committed:
2/3.2

Yandeks committed:
Update main_vArray.cpp

Vipon commented:
Ты немного непонял задание относительно Array и vArray. Array должен быть статическим, иными словами, там не должно быть динамического выделения памяти вообще. vArray наоборот массив переменной длины, иными словами его размер должен быть изменяемым в зависимости от ситуции, если пользователь хочет добавить новый элемент в массив push_back, к примеру, а места нет, то вы должны выделить для него место, если массив сейчас на 100 элементов, а пользователь хочет его уменьшить до 10, то и такая возможность должна быть.

Yandeks committed:
pract

Yandeks committed:
full

Array теперь статический, а vArray динамический. Добавил много функций к vArray

Yandeks committed:
Update main_Array.cpp

Yandeks committed:
Update Array.h

Yandeks committed:
Update main_vArray.cpp

Yandeks committed:
Update Array.h

Yandeks committed:
Update vArray.h

Yandeks committed:
Update vArray.h

Yandeks committed:
Update main_Array.cpp

Yandeks committed:
Update vArray.h

Yandeks committed:
Update vArray.h

Yandeks committed:
Update vArray.h

Yandeks commented:
Я  сделал вещи - все должно работать

Yandeks committed:
Create Theory_8

Yandeks committed:
Create Makefile

Yandeks committed:
Update Makefile

Yandeks committed:
update

Yandeks committed:
Update List_main.cpp

Yandeks committed:
Update main_Array.cpp

Yandeks committed:
Update Makefile

Yandeks committed:
Update main_Array.cpp

Yandeks committed:
Update vArray.h

Yandeks committed:
Update main_vArray.cpp

Vipon commented on Babich_8_home_work/Array.h:
```diff
@@ -0,0 +1,84 @@
+	}
+
+private:
+	T* const table;
```
Массив должен быть статическим.

Vipon commented on Babich_8_home_work/List.h:
```diff
@@ -0,0 +1,87 @@
+#include <iostream>
+
+template <typename T>
+struct Line {
```
Во-первых, название немного странное (необычное вернее), но ладно. Во-вторых, зачем объявление сделано за пределом класса List, если это элемент списка?

Yandeks replied:
Мне нужно так объявить Line, потому что функция DelLine принимает тип Line*, т.е. я использую структуру вне класса.

Vipon replied:
!!!!!

Vipon commented on Babich_8_home_work/List.h:
```diff
@@ -0,0 +1,87 @@
+class List
+{
+public:
+	List(T firstline)
```
Почему нельзя создать пустой список?

Vipon commented on Babich_8_home_work/List.h:
```diff
@@ -0,0 +1,87 @@
+public:
+	List(T firstline)
+	{
+		this->enter = new struct Line<T>;
```
Зачем везде используешь указатель this?

Yandeks replied:
Чтобы было видно, что это private элемент. Без this тоже бы работало.

Vipon commented on Babich_8_home_work/vArray.h:
```diff
@@ -0,0 +1,167 @@
+public:
+	vArray()
+	{
+		this->table = new T[Size];
```
Просто интересно, ты хотя бы ккомпилировал свой код?

Vipon commented on Babich_8_home_work/vArray.h:
```diff
@@ -0,0 +1,167 @@
+class vArray 
+{
+public:
+	vArray()
```
Думается мне, было бы неплохо иметь конструктор, который позволяет сразу создать массив определённого размера.

Vipon commented on Babich_8_home_work/vArray.h:
```diff
@@ -0,0 +1,167 @@
+
+	friend std::istream& operator>>(std::istream& stream, vArray<T>& input)
+	{
+		for (size_t i = 1; i < input.Size + 1; i++)
```
Ахаха, нет, это же динамический массив, а не статический, он должен вместить в себя все элементы из входного потока. При этом нам изначально не известно сколько там элементов. Массив динамический может быть спокойно пустым вообще, его size может быть равен 0, а вот после записи из стандартного потока ввода не нулевым.

Vipon commented on Babich_8_home_work/vArray.h:
```diff
@@ -0,0 +1,167 @@
+		return stream;
+	}
+
+	void push(T data)
```
Такой метод называется push_back, потому есть ещё метод push_forward, который вставляет в начала контейнера.

Vipon commented on Babich_8_home_work/vArray.h:
```diff
@@ -0,0 +1,167 @@
+		return;
+	}
+
+	void state(size_t newSize)
```
Очень странное название метода, посмотри в stl, как подобные методы называются.

Vipon commented on Babich_8_home_work/vArray.h:
```diff
@@ -0,0 +1,167 @@
+
+	void push(T data)
+	{
+		state(this->Size + 1);
```
Странное решение увеличивать размер на +1. Это же ты так каждый раз будешь вызывать выделения памяти.

Yandeks replied:
Это чтобы не было путаницы с  различием в размере массива и с номером последнего элемента. К тому же, чтобы эффективно работать с памятью, можно использовать set_size и [].

Yandeks replied:
а нет, путаница уже есть из-за set_size

Vipon commented on Babich_8_home_work/vArray.h:
```diff
@@ -0,0 +1,167 @@
+	void push(T data)
+	{
+		state(this->Size + 1);
+		this->table[this->Size -1] = data;
```
А если в массиве элементов меньше чем Size?

Vipon commented on Babich_8_home_work/vArray.h:
```diff
@@ -0,0 +1,167 @@
+
+	vArray& operator=(const vArray<T>& dupl)
+	{
+		memcpy(this->table, dupl.table, sizeof(T)*dupl.Size);
```
А у тебя места-то в массиве хватит?

Yandeks committed:
>>

Yandeks committed:
Update Array.h

Yandeks committed:
Update List.h

Yandeks committed:
Update main_Array.cpp

Yandeks committed:
Update vArray.h

Yandeks committed:
Update Array.h

Yandeks committed:
Update vArray.h

Yandeks committed:
Update List.h

Yandeks committed:
Update vArray.h

Yandeks committed:
Update vArray.h

Yandeks committed:
Update vArray.h

Yandeks committed:
Update vArray.h

Vipon commented on Babich_8_home_work/Array.h:
```diff
@@ -0,0 +1,85 @@
+	}
+
+private:
+	T* table;
```
Память под массив выделяется на куче, а у тебя должен был быть статический массив.

Yandeks committed:
Update vArray.h

Yandeks committed:
Update vArray.h

Yandeks committed:
Update vArray.h

Yandeks committed:
Update vArray.h

Yandeks committed:
Update Array.h

Yandeks committed:
Update Array.h

Yandeks committed:
Update Array.h

Yandeks committed:
Update main_Array.cpp

Yandeks committed:
Update vArray.h

Yandeks committed:
Update Array.h

Yandeks committed:
Update vArray.h

Vipon commented on Babich_8_home_work/Array.h:
```diff
@@ -0,0 +1,82 @@
+
+	friend std::istream& operator>>(std::istream& stream, Array<T, Size>& input)
+	{
+		for (size_t i = 1; i < Size+1; i++)
```
Что будет, если на вход подано 2 элемента, а массив на 100?

Yandeks committed:
Update List.h

Yandeks committed:
Update List_main.cpp

Yandeks committed:
Update List.h

Yandeks committed:
Update List.h

Yandeks committed:
Update List.h

Yandeks committed:
Update List.h

Yandeks committed:
Update List_main.cpp

Yandeks committed:
Update List.h

Yandeks committed:
Update List.h

Yandeks committed:
Update List.h

Yandeks committed:
Update List.h

Yandeks committed:
Update Array.h

Yandeks committed:
Update Array.h

Yandeks committed:
Update Array.h

Yandeks committed:
Update Array.h

Yandeks committed:
Update Array.h

Yandeks committed:
Update Array.h

Yandeks committed:
Update Array.h

Vipon commented on Babich_8_home_work/Array.h:
```diff
@@ -0,0 +1,99 @@
+				return stream;
+			}
+			
+			input.table[i] = (T)stod(buf);
```
Ты прочитал строку, взял первое число из него и всё остальное выкинул? И условие остановки довольно странное, почему пробел?

Yandeks replied:
stod берет все число, а не первую цифру. В новом цикле buf обновляется и там будет другое число. То есть, тут не все в одну строку вводится, а через enter. А в чем проблема пробела в таком методе?

Vipon commented on Babich_8_home_work/Array.h:
```diff
@@ -0,0 +1,99 @@
+	friend std::istream& operator>>(std::istream& stream, Array<T, Size>& input)
+	{
+		std::string buf = "oh shit";
+		for(int i = 0; i< (input.last_pos-1); i++)
```
Вообще ни разу не выполнится, потому что input.last_pos-1 = -1

Yandeks replied:
я хотел сделать ввод до последнего текущего элемента, но это все странно как-то - сейчас переделаю

Vipon commented on Babich_8_home_work/Array.h:
```diff
@@ -0,0 +1,99 @@
+	{
+		if (pos < (last_pos-1))
+		{
+			return table[pos - 1];
```
Вообще ни разу правильно не отработает, потому что input.last_pos-1 = -1

Yandeks replied:
все же верно? Если last_pos =0, то массив пуст и брать нечего

Yandeks committed:
Update Array.h

Yandeks committed:
Update Array.h

Vipon commented on Babich_8_home_work/List.h:
```diff
@@ -0,0 +1,127 @@
+		return stream;
+	}
+
+	 T& operator[](size_t pos)
```
Это список, а не массив.

Vipon commented on Babich_8_home_work/List.h:
```diff
@@ -0,0 +1,127 @@
+	 T& operator[](size_t pos)
+	{
+		if (pos > length)
+			return nullptr;
```
Где исключения.

Vipon commented on Babich_8_home_work/List.h:
```diff
@@ -0,0 +1,127 @@
+		}
+		delete[] buf1;
+	}
+	friend std::istream& operator>>(std::istream& stream, List<T>& input)
```
Как заполнить пустой список?

Yandeks committed:
Update List.h

Yandeks committed:
Update Array.h

Yandeks committed:
Update List.h

Yandeks committed:
Update List.h

Yandeks committed:
Update List.h

Yandeks committed:
Update List.h

Yandeks committed:
Update List.h

Yandeks committed:
Update List.h

Yandeks committed:
Update List.h

Yandeks committed:
Update Theory_8

Yandeks committed:
Update Array.h
